### DISCUSSION TITLE
Suppress fixer rules using @nofixer annotation

### DISCUSSION THREAD

**Discussion by anton-kotik** [PHP-CS-Fixer/PHP-CS-Fixer#7932] [Feature ideas]
### Idea

Suppression of any rule with an annotation:

```php
/** @nofixer [fixer_rule_name] [Optional comment] */
```

```php
/** @nofixer [ClassNameFixer] [Optional comment] */
```

The idea taken from the [inspections suppression](https://www.jetbrains.com/help/phpstorm/disabling-and-enabling-inspections.html#suppress-inspections) IDE feature aka `/** @noinspection */` comments.

To suppress the rule, you need to add the `/** @nofixer */` annotation before the selected piece of code (entire file, class, method, property, or statement). For example, if you suppress a rule for a class, add the annotation before this class. When a rule is suppressed, php-cs-fixer doesn't fix this rule problem found in the specific piece of code.

### Example
```php
<?php

/** @nofixer single_blank_line_at_eof */
/** @nofixer declare_strict_types This will be fixed in X-205 task */
```

UpperCamelCase looks similar to `@noinspection` style and also could be supported:
```php
<?php

/** @nofixer SingleBlankLineAtEofFixer */
/** @nofixer DeclareStrictTypesFixer This will be fixed in X-205 task */
/** @noinspection PhpPropertyOnlyWrittenInspection */
```

### Primitive implementation

We are using something like this code for implementing this feature in our projects (works only for the entire file suppression):

```php
public function isCandidate(Tokens $tokens): bool {
    return !$this->isFixerSuppressed($tokens);
}

/**
 * Detects rules disabled on file level.
 * If before namespace / use there is {@nofixer SomeFixer} or {@nofixer} comment,
 * disable relevant rules.
 */
private function isFixerSuppressed(Tokens $tokens): bool {
    if (!$tokens->isTokenKindFound(T_DOC_COMMENT)) {
        return false;
    }

    $nameOrClass = '(' . $this->getName() . '|' . basename(str_replace('\\', '/', static::class)) . ')';
    foreach ($tokens as $token) {
        // Variants:
        /** @nofixer */
        /** @nofixer rule_name */
        /** @nofixer SomeFixer */
        /** @nofixer SomeFixer X-205 */
        if (
            $token->getId() === T_DOC_COMMENT
            && preg_match('/^\\/\*\* @nofixer( ' . $nameOrClass . '( .+)?)? \*\\/$/', $token->getContent())
        ) {
            return true;
        }

        // We don't look beyond these tokens:
        if ($token->equalsAny([[T_NAMESPACE], [T_USE], [T_CLASS], [T_VARIABLE], [T_CLOSE_TAG], '{'])) {
            return false;
        }
    }

    return false;
}
```

**Comment by Wirone** [comment]
Hi @anton-kotik, thanks for detailed proposal! FYI: it was already discussed in #3871 and even there was an attempt to implement it in #7569 🙂. There's also #4512, so a lot of stuff to get familiar with 😅.

**Comment by keradus** [comment]
I believe this is covered by https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules_exceptions.rst
